fs/zbd_zenfs.cc

```diff
@@ -917,6 +918,91 @@ Zone *ZonedBlockDevice::AllocateZone(Env::WriteLifeTimeHint file_lifetime, bool
   return allocated_zone;
 }
 
+Zone *ZonedBlockDevice::AllocateDataZone(Env::WriteLifeTimeHint file_lifetime, bool is_wal) {
+  // Ignore lifetime hint for now
+  (void)file_lifetime;
+
+  Zone* ret = nullptr;
+  if (is_wal) {
+    fg_request_++;
+    do {
+      {
+        std::unique_lock<std::mutex> lk(active_zone_vec_mtx_);
+        ret = GetDataZoneLocked(0);  
+      }
+      if (ret == nullptr) {
+        // Resources not enough, release lock for background worker and waiting.
+        std::this_thread::yield();
+      }
+    } while (ret == nullptr);
+    fg_request_--;
+  } else {
+    // Normal file
+    int request = 0;
+    do {
+      // Test if there is high priority requsets.
+      request = fg_request_.load();
+      if (request == 0) {
+        // Fetch one when no contention occurs.
+        std::unique_lock<std::mutex> lk(active_zone_vec_mtx_);
+        ret = GetDataZoneLocked(reserved_active_zones);
+      } else {
+        std::this_thread::yield();
+      }
+    } while (ret == nullptr);
+  }
+  return ret;
+}
+
+Zone *ZonedBlockDevice::GetDataZoneLocked(int start) {
+  for (int i = start; i < active_zones_vec_.size(); ++i) {
+    if (active_zones_vec_[i]->state_ == kActive) {
+      active_zones_vec_[i]->state_ = kOccupied;
+      return active_zones_vec_[i];
+    }
+  }
+  return nullptr;
+}
+
+void ZonedBlockDevice::BgRefillActiveZones() {
+  data_worker_->SubmitJob(
+    [&]() {
+      for (int i = 0; i < active_zones_vec_.size(); ++i) {
+        if (active_zones_vec_[i]->state_ == kReadOnly) {
```

Comment:
Smells like a data racing here, shall we put this `if` condition inside the lock?

Reply:
The caller should hold the lock, I might need to illustrate this with comments, maybe add a "locked" suffix to the function also.